R/summaries.R

```diff
@@ -87,3 +105,231 @@ setMethod("getTopTaxa", signature = c(x = "SummarizedExperiment"),
+                }
+            }
+        # Adds dominant taxas to colData
+        tmp <- dominantTaxa(x, abund_values, rank, name)
+        # Gets an overview
+        overview <- .get_overview(tmp, group, name)
+        overview
+  }
+)
+
+################################ HELP FUNCTIONS ################################
+
+#' @importFrom S4Vectors as.data.frame
+#' @importFrom dplyr n desc tally group_by arrange mutate
+.get_overview <- function(x, group, name){
+    # Creates a tibble df that contains dominant taxa and number of times that
+    # they present in samples and relative portion of samples where they
+    # present.
+    if (is.null(group)) {
+        name <- sym(name)
+        overview <- as.data.frame(colData(x)) %>%
+            group_by(!!name) %>%
+            tally() %>%
+            mutate(
+                rel.freq = round(100 * n / sum(n), 1),
+                rel.freq.pct = paste0(round(100 * n / sum(n), 0), "%")
+            ) %>%
+            arrange(desc(n))
+    } else {
+        group <- sym(group)
+        name <- sym(name)
+
+        overview <- as.data.frame(colData(x)) %>%
+            group_by(!!group, !!name) %>%
+            tally() %>%
+            mutate(
+                rel.freq = round(100 * n / sum(n), 1),
+                rel.freq.pct = paste0(round(100 * n / sum(n), 0), "%")
+            ) %>%
+            arrange(desc(n))
+    }
+    return(overview)
+}
+
+#' @rdname summaries-basic
+#'
+#' @param x A
+#'  \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}} object.
+#'
+#' @param abund_values a \code{character} value to select an
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{assayNames}}
+#'   By default it expects count data.
+#'
+#' @details
+#' The \code{summarizeSE} will return a summary of counts for all samples and
+#' features in
+#' \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#' object.
+#'
+#' @return
+#' The \code{summarizeSE} returns a list with two \code{tibble}s
+#'
+#' @seealso
+#' \code{\link[scuttle:perCellQCMetrics]{perCellQCMetrics}},
+#' \code{\link[scuttle:perFeatureQCMetrics]{perFeatureQCMetrics}},
+#' \code{\link[scuttle:addPerCellQC]{addPerCellQC}},
+#' \code{\link[scuttle:addPerFeatureQC]{addPerFeatureQC}},
+#' \code{\link[scuttle:quickPerCellQC]{quickPerCellQC}}
+#'
+#' @export
+setGeneric("summarizeSE",
+           signature = c("x"),
+           function(x, ...)
+             standardGeneric("summarizeSE")
+)
+
+#' @rdname summaries-basic
+#' @export
+setMethod("summarizeSE",
+          signature = c(x = "SummarizedExperiment"),
+          function(x, abund_values = "counts"){
+            .check_rel_neg(x, abund_values)
+            sample.summary <- .get_summary_col_data(x, abund_values)
+            feature.summary <- .get_summary_rowl_data(x, abund_values)
+            return(list("samples" = sample.summary, "features" = feature.summary))
+          }
+)
```

Comment:
I got a problem with the name `summarizeSE`.

Two things I would like to throw into the mix:
1. The object class type is in the method name. This is redundant, since we are using S4 dispatches all around. Therefore, this is already done for us.
2. Is this really a summary or is a quality control function implemented here?

From these two things, I am thinking about the verb for this function. If it is really a summary, one could just the verb `summary`, for which the generic is defined in base R. If it is about QC maybe the naming scheme used elsewhere might be better followed, e.g. `perTaxaQCMetrics`/`addPerTaxaQC` and `perSampleQCMetrics`/`addPerSampleQC`

Reply:
These are more like stats for overall denoised sequences and inferred ASVs. QC would include more information and covered by `perCellMetrics` and `perFeatureMetrics`. Would `seqeunceStats()` be a more appropriate name? 
For `perFeatureMetrics` I was thinking of expanding it to include other information shown in **2b** eg., here https://github.com/microbiome/mia/issues/69 . This will be in qc-functions.rd file.

Reply:
So there are two aspects these functions need to provide as far as I understand it now:
- more then QC (so the output of `perFeatureQCMetrics` does not suffice as explained by example figure in the issue)
- but the layout of the result should be along the dimension of the input

Taking this into account, the functionality would by described by functions named `perTaxaSummary` and `perSampleSummary`? (plus the additional `add*` wrappers) This would marry the idea of a summary with the granuality provided by other "Metric" functions, provide the context what is summarized and be granular enough, so that either summary could be opbtained individually.

I am not sure `seqenceStats` describes the functionality properly. What is `sequence` in this context? What would expected from a function named `seqenceStats` generally? I would guess something like GC content.

Reply:
Yes. The `perTaxaSummary` will at least be a separate function together with add* to `rowData`. At the moment, I am still not sure what additional information can be provided by `perSampleSummary` that `perCellQCMetrics` from _scuttle_ doesn't provide. 

The `summarizeSE` returns the following output as a list of `tibbles`.
![image](https://user-images.githubusercontent.com/16476233/120204200-d7353180-c228-11eb-91d4-ca1a3f4ea165.png)
My first thought was that these are denoised sequencing stats. It should be more like `sequencingStats()` and not `sequenceStats()` but still it may not be ideal :(

Reply:
`sequencingStats` implies a specific technique was used. Would this function then not be usable for array data?

`inputStats` would then be more appropriate, but it is of course logical, that the stats are computed on the input. This leaves just `stats`, which is another way of saying `summary`. With this I have brought it back to my first somment: Maybe just using the base R generic `summary` is suitable best.

Reply:
Okay. So I need to do it like this?
```
setGeneric("summary",
           signature = c("x"),
           function(x, ...)
             standardGeneric("summary")
)

#' @rdname summaries-basic
#' @export
setMethod("summary",
          signature = c(x = "SummarizedExperiment"),
          function(x, abund_values = "counts"){
            .check_rel_neg(x, abund_values)
            sample.summary <- .get_summary_col_data(x, abund_values)
            feature.summary <- .get_summary_rowl_data(x, abund_values)
            return(list("samples" = sample.summary, "features" = feature.summary))
          }
)
```

Reply:
The first part you don't need, since the generic already exists. 

I am not sure whether you really need to add the following roxygen tag to the chunk preceeding `setMethod`:
```
#' @importFrom base summary
```
... since the base package is always present. Try and see if roxygen or a check complains.

Reply:
Removed `setGeneric` and added `#' @importFrom base summary`. I get this error

```
Error in matchSignature(signature, fdef, where) : 
  there are named arguments (‘x’) in the method signature that are missing from the generic signature, for function ‘summary’
Error in unregister_namespace(package) : 
  mia is not a registered namespace.
```

Reply:
```r
getGeneric("summary")
standardGeneric for "summary" defined from package "base"

function (object, ...) 
standardGeneric("summary")
<environment: 0x0000019d241e6628>
Methods may be defined for arguments: object
Use  showMethods("summary")  for currently available ones.
```

The first input has to be name `object` and the dispatch also has to be set to the argument `object`. Like this:

```r
setMethod("summary", signature = c(object = "SummarizedExperiment"),
           ...)
```

Reply:
Now working atleast roxygen not showing error.